pennylane/ops/op_math/prod.py

```diff
@@ -0,0 +1,210 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains the implementation of the Prod class which contains logic for
+computing the product between operations.
+"""
+from copy import copy
+from functools import reduce
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import math
+from pennylane.operation import Operator, expand_matrix
+
+
+def prod(*ops, do_queue=True, id=None):
+    """Represent the tensor product (or matrix product) between operators."""
+    return Prod(*ops, do_queue=do_queue, id=id)
+
+
+def _prod(
+    mats_gen,
+):  # TODO: current multiplication always expands mats, this is not always necessary
+    """Multiply matrices together"""
+    res = reduce(
+        math.dot, mats_gen
+    )  # Inefficient method (should group by wires like in tensor class)
+    return res
+
+
+class Prod(Operator):
+    """Arithmetic operator subclass representing the scalar product of an
+    operator with the given scalar."""
+
+    _name = "Prod"
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+
+    def __init__(self, *factors, do_queue=True, id=None):  # pylint: disable=super-init-not-called
+        """Initialize a Prod instance"""
+        self._id = id
+        self.queue_idx = None
+
+        if len(factors) < 2:
+            raise ValueError(f"Require at least two operators to multiply; got {len(factors)}")
+
+        self.factors = factors
+        self._wires = qml.wires.Wires.all_wires([s.wires for s in self.factors])
+
+        if do_queue:
+            self.queue()
+
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        return " @ ".join([f"{f}" for f in self.factors])
+
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.factors = tuple(f.__copy__() for f in self.factors)
+        copied_op.data = self.data.copy()  # copies the combined parameters
+
+        for attr, value in vars(self).items():
+            if attr not in {"data", "factors"}:
+                setattr(copied_op, attr, value)
+
+        return copied_op
```

Comment:
Why can't we do: `return Prod(tuple(copy(f) for f in self.factors))`? Is there any attribute that is dynamically changed?

Reply:
Other similar symbolic ops have needed this when they have more information that has to be copied over.  This is just mimicking that same behaviour.  I think doing this will also be more future-proof, since it will still work if we start adding in more hyperparameters or cached properties.